org.eclipse.xtext.util/src/org/eclipse/xtext/util/MergeableManifest2.java

```diff
@@ -671,4 +645,130 @@ public boolean equals(Object obj) {
+			if (list.isEmpty()) {
+				list.add(newBundle);
+				return;
+			}
+			boolean merged = false;
+			for (int i = 0; i < list.size(); i++) {
+				Bundle oldBundle = list.get(i);
+				if (oldBundle.hasSameName(newBundle)) {
+					String oldBundleName = oldBundle.getName();
+					String oldBundleVersion = oldBundle.getVersion();
+					String oldBundleSuffix = oldBundle.getSuffix();
+					String bundleVersion = oldBundleVersion == null ? newBundle.getVersion() : oldBundleVersion;
+					merged = true;
+					if (oldBundleSuffix == null) {
+						if (bundleVersion != null) {
+							list.set(i, Bundle.fromNameVersion(oldBundleName, bundleVersion));
+						}
+					} else {
+						if (bundleVersion != null) {
+							if (oldBundleVersion == null) {
+								list.set(i,
+										Bundle.fromNameVersionSuffix(oldBundleName, bundleVersion, oldBundleSuffix));
+							}
+						}
+					}
+				}
+			}
+			if (!merged)
+				list.add(newBundle);
+		}
+
+		// Output is used by algorithm ... do not change for debugging purposes!
+		@Override
+		public String toString() {
+			String separator = "";
+			StringBuilder result = new StringBuilder("");
+			for (Bundle bundle : list) {
+				result.append(separator).append(bundle);
+				separator = ",";
+			}
+			return result.toString();
+		}
+
+	}
+
+	private static class Bundle {
+		private final String input;
+		private final List<String> split;
+
+		public static Bundle fromInput(String input) {
+			return new Bundle(input);
+		}
+
+		public static Bundle fromNameVersion(String name, String version) {
+			return new Bundle(name + ";bundle-version=\"" + version + "\"");
+		}
+
+		public static Bundle fromNameVersionSuffix(String name, String version, String suffix) {
+			return new Bundle(name + ";bundle-version=\"" + version + "\";" + suffix);
+		}
+
+		private Bundle(String input) {
+			this.input = input;
+			if (this.input.contains(";"))
+				this.split = Collections.unmodifiableList(splitAtCharHonorQuoting(this.input, ';'));
+			else
+				this.split = Collections.singletonList(this.input);
+		}
+
+		public boolean hasSameName(Bundle other) {
+			return Objects.equals(getName(), other.getName());
+		}
+
+		public String getName() {
+			return split.get(0);
+		}
+
+		public String getSuffix() {
+			return split.size() > 1 ? split.subList(1, split.size()).stream().reduce((a, b) -> a + ";" + b).get()
+					: null;
+		}
+
+		public String getVersion() {
+			for (int n = 1; n < split.size(); n++) {
+				String part = split.get(n);
+				if (part.contains("bundle-version=")) {
+					int startIndex = part.indexOf("bundle-version=") + "bundle-version=".length();
+					if (part.charAt(startIndex) == '"') {
+						return part.substring(startIndex + 1, part.indexOf("\"", startIndex + 1));
+					} else {
+						return part.substring(startIndex);
+					}
+				}
+			}
+			return null;
+		}
+
+		// Output is used by algorithm ... do not change for debugging purposes!
+		@Override
+		public String toString() {
```

Comment:
Maybe add a dedicated method and let toString call that one and make the algo based on that method?